### ISSUE TITLE
Unifying Dongle: Marking firmware complete takes 24 Seconds, doesn't boot into firmware mode

### ISSUE THREAD

**Issue by mkay100** [RFStorm/mousejack#42]
First of all: Thank you for the amazing work.

When flashing my unifying dongle (all preparations ahead no errors) i get the following results:
2019-05-07 07:52:11.419]  Mark firmware update as completed
Traceback (most recent call last):
  File "./prog/usb-flasher/logitech-usb-flash.py", line 61, in <module>
    response = dongle.send_command(0x21, 0x09, 0x0200, 0x0000, "\x20\x00\x00\x01\x02" + "\x00"*27)
  File "/root/nrf/nrf-research-firmware/prog/usb-flasher/unifying.py", line 180, in send_command
    response = self.dongle.read(ep, 32, timeout=timeout)
  File "/usr/local/lib/python2.7/dist-packages/usb/core.py", line 988, in read
    self.__get_timeout(timeout))
  File "/usr/local/lib/python2.7/dist-packages/usb/backend/libusb1.py", line 851, in intr_read
    timeout)
  File "/usr/local/lib/python2.7/dist-packages/usb/backend/libusb1.py", line 936, in __read
    _check(retval)
  File "/usr/local/lib/python2.7/dist-packages/usb/backend/libusb1.py", line 595, in _check
    raise USBError(_strerror(ret), ret, _libusb_errno[ret])
usb.core.USBError: [Errno 110] Operation timed out
Makefile:33: recipe for target 'logitech_install' failed
make: *** [logitech_install] Error 1

When i alter the usb_timeout in unifying.py from 2500 to 32500 it seems to work:

[2019-05-07 07:54:46.032]  Mark firmware update as completed
[2019-05-07 07:55:09.883]  24:00:00:00:00:6B:32:FF:FF:FF:FF:FF:FF:FF:32:FF:FF:FF:FF:FF:FF:6F:B7:FF:41:A0:B4:15:FB:7E:79:B3
[2019-05-07 07:55:09.884]  Restarting dongle into research firmware mode
[2019-05-07 07:55:09.886]  70:00:00:00:00:6B:32:FF:FF:FF:FF:FF:FF:FF:32:FF:FF:FF:FF:FF:FF:6F:B7:FF:41:A0:B4:15:FB:7E:79:B3

But the device still shows up as:
Bus 001 Device 024: ID 046d:aaaa Logitech, Inc. 

Any advice would be appreciated.

**Comment by mkay100** [comment]
Additional info on the Device:

Unifying [bootloader-nordic]
  DeviceId:             8f98944e4c0d22c8a5b5b250eda4d3f037fc6fa7
  Guid:                 9d131a0c-a606-580f-8eda-80587250b8d6
  Summary:              A miniaturised USB wireless receiver (bootloader)
  Plugin:               unifying
  Flags:                updatable|supported|registered|is-bootloader
  Vendor:               Logitech
  VendorId:             USB:0x046D
  Version:              RQR00.00_B0000
  VersionBootloader:    BOT01.04_B0016
  Icon:                 preferences-desktop-keyboard
  Created:              2019-05-07

**Comment by 4dvn** [comment]
I sovled this error by this step:
Cd nrf (renamed folder after git clone from git)
Cd prog
Cd usb-flasher 
(You need to copy 4 files compiled from bin folder to this folder)
Python logitech-usb-flash.py dongle_formated.bin dongle_formated.ihx
==> done
Just retype lsusb to see the  @result:

**Comment by 4dvn** [comment]
![98D2E12E-5D7E-4B9C-BAE4-D7F79253DC9B](https://user-images.githubusercontent.com/46510477/61822926-f6bcdf80-ae84-11e9-911e-a6a37b163845.png)

**Comment by mame82** [comment]
The last operation in Marc's USB flasher writes a single byte to flash address 0×0000.

This write triggers the CRC check for the whole firmware blob, and thus takes some seconds.

On dongles with latest bootloader, this additionally would trigger a signature check. The signature check would take longer and fail in any case, as the firmware just has no valid 256byte signature (which has to be send to the receiver with additional commands). Trying to flash a Nordic dongle which has a signed Bootloader, would wipe the firmware, although flashing fails. In such a case the dongle would be stuck in bootloader mode.

For Nordic dongles, signed bootloader start with version >=BOT01.04

So make sure you have <=BOT01.03 before trying to flash

**Comment by ltsalvatore** [comment]
damn.. where is my post from a couple of days ago?

**Comment by seeklakhan** [comment]
@mame82 any way we can update bootloader if we are stuck in booterloader mode?

**Comment by FyiurAmron** [comment]
@seeklakhan for you and posteriority (this is my own research; might be wrong in some details, but it works as a whole; more experienced people would have to chime in; disclaimer: I'm both Windows and Linux user, so this covers both):
YES. If you want the solution, scroll to the bottom. If you want some reading, bear with me.
### problem ###
This script doesn't handle the secure flashing at all. As a result, it flashes the memory part that is accessible (see your infopage offset 0x20 for how much that is; in general, "nRF24LU1+ Product Specification" PDF from Nordic Semi is a blessing here if you want to know how to hack this device). Also, it doesn't do the verification etc. that's needed afterwards, so you're basically getting a soft-bricked unit, locked in bootloader because it detects unauthorized write and that's it.
This tool or any of the ages-old Bastille scripts won't allow you to easily fix it, because stock firmware will usually be slightly off from your original one (edge case where there were FFs instead of 00s present in the prior data - leftovers from raw flash memory) screwing with the checksum. Also, it won't handle the verification properly anyway AFAIU. Theoretically, you should be able to write the _exact_ dump of your old firmware into that region, but 1. that glitched for me, inserting some write errors for some reason, 2. still that doesn't finish the procedure fully AFAIU.
### solution ###
Well, obviously you need a way to re-flash just like a pro, right? So, obviously, Logitech soft should be able to do that, right?
...
nooooope. New Logi+ mobile-like crap app won't even detect a bootloader-state device properly. Older Unifying Software _will_ detect it, but then will fail on updating.

AFAIK, the only tool that has a full, verified and industry-standard flashing procedure is, unsurprisingly, `fwupd`. Lo and behold, there are versions for both Linux and Windows, should be easy, right?
NOOOOPE :D

Firstly, Windows version lacks the `logitech_hidpp` plugin which is needed for it, so, yeah, build from scratch with this plugin re-enabled, ages spent on all the ugly corner cases, but then...

IT DOESN'T WORK XD

`fwupd` segfaults and throws random errors.

"Wow, it's bricked, prolly a lost case."... no, it's just a bug in 2.x `fwupd` for this particular scenario, flashing secured hex files :D https://github.com/fwupd/fwupd/issues/10682

By now you know it's a lost case, because Windows builds are really rare for 1.x `fwupd` (e.g. 1.9.34 doesn't have it at all), and they don't have the plugins anyway, and it's archaeology by that point. So, WSL, and simple `apt install`?

Sure, as long as your distro still has 1.9.34 available, i.e. 24 LTS max :D Otherwise, you'll need to download

```
fwupd_1.9.34-0ubuntu1~24.04.1_amd64.deb
libfwupd2_1.9.34-0ubuntu1~24.04.1_amd64.deb
libgusb2_0.4.8-1build2_amd64.deb
```

from a mirror of your choice, and install them manually as well. Alternatively - download from git, build from source, or try finding valid binaries for your distro. Simple. Run... Only, it won't detect your dongle at all?! ...

And, then you'll realize you need `usbipd` (if you don't have it already) on your Windows side, so you install it, do the bind/attach dance, and try again...

And it says "not upgradable?!"...

Yes, you need to `fwupd refresh`, because it doesn't have the firmware dump itself yet :D

OK, now it says it's upgradable, so it will work, right? You run it and...

and the flashing fails almost immediately. "It's bricked, lost case, all this time wasted..."
No. It's just `usbipd` not handling the USB refresh. The flasher pings the device to switch mode, USB is re-enumerated, and `usbipd` loses the attachment.

So, then you find the `usbpid attach --auto-attach` (`-a` in short) trick, aaaand...

Goes slowly, then gets stuck for about half a minute... then some strange error? Surely it's bri...

Then you move the mouse with your elbow by accident, and the cursor moves. :D

YES, IT WORKS. (8 hours later :DDD)

---
tl;dr use `fwupd` v1.9.34 on Linux (do a refresh first to get newest fw). On WSL, use `usbipd` with `--auto-attach`. Done.

**Comment by mame82** [comment]
I wrote a tool to deal with Logitech receivers of different kinds. It also supported firmware flashing (for both, signed and unsigned bootloader's).

Link: https://github.com/RoganDawes/munifying

@seeklakhan updating the bootloader itself is not possible via USB, as it is basically the remaining piece of software handling the USB communication. If you want to do this (bootloader exchange), you have to go physical and wire up the dongle to a hardware debugger (tap chip pins).
